### DISCUSSION TITLE
System.Exception: Could not find metadata for 'Newtonsoft.Json.Linq.JObject' occur when using Sink Task

### DISCUSSION THREAD

**Discussion by GustavXie** [ravendb/ravendb#17891] [Q&A]
Recently I am using RavenDB Queue Sink: RabbitMQ feature to make RavenDB as a RabbitMQ message consumer. 
So I did that based on this [doc ](https://ravendb.net/docs/article-page/6.0/csharp/server/ongoing-tasks/queue-sink/rabbit-mq-queue-sink#queue-sink-rabbitmq)

However when I try to test it sending a test message to RabbitMQ, it will have 50% to have error as following:

System.Exception: Could not find metadata for 'Newtonsoft.Json.Linq.JObject'.
Ensure the following:
1. 'Newtonsoft.Json.Linq.JObject' is included in initial scanning.
2. 'Newtonsoft.Json.Linq.JObject' implements either 'IMessage', 'IEvent' or 'ICommand' or alternatively, if you don't want to implement an interface, you can use 'Unobtrusive Mode'. 

But if I send the same message again, it will not have any issue and create the document in RavenDB successfully. For this case, how could I solve this? Thanks

**Comment by arekpalinski** [comment]
Can you please share the full exception message including the stacktrace? That could tell us more about your issue.

CC @djordjedjukic

**Reply by GustavXie** [reply]
```
message_id:	f3db4e0d-c187-4e7d-9b1a-f2447004c5ba
delivery_mode:	2
headers:	
$.diagnostics.hostdisplayname:	c599caa9d9cb
$.diagnostics.hostid:	48e1618bf59f983beb9a6ab1f8bee5cd
NServiceBus.ExceptionInfo.Data.Message ID:	f3db4e0d-c187-4e7d-9b1a-f2447004c5ba
NServiceBus.ExceptionInfo.ExceptionType:	NServiceBus.MessageDeserializationException
NServiceBus.ExceptionInfo.HelpLink:	undefined
NServiceBus.ExceptionInfo.InnerExceptionType:	System.Exception
NServiceBus.ExceptionInfo.Message:	An error occurred while attempting to extract logical messages from incoming physical message f3db4e0d-c187-4e7d-9b1a-f2447004c5ba
NServiceBus.ExceptionInfo.Source:	NServiceBus.Core
NServiceBus.ExceptionInfo.StackTrace:	NServiceBus.MessageDeserializationException: An error occurred while attempting to extract logical messages from incoming physical message f3db4e0d-c187-4e7d-9b1a-f2447004c5ba
---> System.Exception: Could not find metadata for 'Newtonsoft.Json.Linq.JObject'.
Ensure the following:
1. 'Newtonsoft.Json.Linq.JObject' is included in initial scanning.
2. 'Newtonsoft.Json.Linq.JObject' implements either 'IMessage', 'IEvent' or 'ICommand' or alternatively, if you don't want to implement an interface, you can use 'Unobtrusive Mode'.
at NServiceBus.Unicast.Messages.MessageMetadataRegistry.GetMessageMetadata(Type messageType) in /_/src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs:line 46
at NServiceBus.Pipeline.LogicalMessageFactory.Create(Type messageType, Object message) in /_/src/NServiceBus.Core/Pipeline/Incoming/LogicalMessageFactory.cs:line 51
at NServiceBus.DeserializeMessageConnector.Extract(IncomingMessage physicalMessage) in /_/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs:line 129
at NServiceBus.DeserializeMessageConnector.ExtractWithExceptionHandling(IncomingMessage message) in /_/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs:line 48
--- End of inner exception stack trace ---
at NServiceBus.DeserializeMessageConnector.ExtractWithExceptionHandling(IncomingMessage message) in /_/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs:line 52
at NServiceBus.DeserializeMessageConnector.Invoke(IIncomingPhysicalMessageContext context, Func`2 stage) in /_/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs:line 30
at NServiceBus.UnitOfWorkBehavior.InvokeUnitsOfWork(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/UnitOfWork/UnitOfWorkBehavior.cs:line 40
at NServiceBus.MutateIncomingTransportMessageBehavior.InvokeIncomingTransportMessagesMutators(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/MessageMutators/MutateTransportMessage/MutateIncomingTransportMessageBehavior.cs:line 59
at NServiceBus.ProcessingStatisticsBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Performance/Statistics/ProcessingStatisticsBehavior.cs:line 25
at NServiceBus.TransportReceiveToPhysicalMessageConnector.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs:line 37
at NServiceBus.RetryAcknowledgementBehavior.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/ServicePlatform/Retries/RetryAcknowledgementBehavior.cs:line 25
at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 35
at NServiceBus.TransportReceiver.InvokePipeline(MessageContext c) in /_/src/NServiceBus.Core/Transports/TransportReceiver.cs:line 58
at NServiceBus.TransportReceiver.InvokePipeline(MessageContext c) in /_/src/NServiceBus.Core/Transports/TransportReceiver.cs:line 64
at NServiceBus.Transport.RabbitMQ.MessagePump.Process(EventingBasicConsumer consumer, BasicDeliverEventArgs message, Byte[] messageBody) in /_/src/NServiceBus.Transport.RabbitMQ/Receiving/MessagePump.cs:line 368
NServiceBus.FailedQ:	FraudDetection.input
NServiceBus.MessageId:	f3db4e0d-c187-4e7d-9b1a-f2447004c5ba
NServiceBus.NonDurableMessage:	False
NServiceBus.ProcessingEndpoint:	FraudDetection.Api
NServiceBus.ProcessingMachine:	c599caa9d9cb
NServiceBus.TimeOfFailure:	2023-12-19 17:00:03:662365 Z
NServiceBus.Transport.RabbitMQ.ConfirmationId:	1
content_type:	application/octet-stream

Encoding: string
{

    "EventType":"UpdateCreditCard",

	"SessionId":"abea25c5-43b5-4a44-b30d-3762be5b159d",

	"TimeStamp":"2023-12-14 15:30:06:300669 Z",

	"UserId":"13ce6629-40c2-4fec-b05c-a3b79b6f62ee",

	"Values":{}

}

**Reply by arekpalinski** [reply]
This is an exception from NServiceBus. Do you get any exception in RavenDB? I'm afraid I don't follow what and where the issue is. Can you please try to elaborate more about it.

**Reply by ayende** [reply]
I believe that the issue is that he is sending the data from RavenDB to RabbitMQ and trying to read that in NServiceBus.
And getting an error because the format isn't as expected by NSB?

**Reply by ayende** [reply]
I'm **guessing** that he is missing a `$type` in how the message is formatted, so json.net will know how to properly figure out what type to deserialize it to.

**Reply by GustavXie** [reply]
I did not see any RavenDB error from the log, but the thing is even I do exactly the same thing as this [doc](https://ravendb.net/docs/article-page/6.0/csharp/server/ongoing-tasks/queue-sink/rabbit-mq-queue-sink#queue-sink-rabbitmq) from the website and test it, I will have the same weird issue that 50% chance to have error as System.Exception: Could not find metadata for 'Newtonsoft.Json.Linq.JObject'. Then If I submit the same test message again it will not any issue which does not make sense.

So if I need to make the sink task work without `Could not find metadata error`, do I need any extra setting for NserviceBus to handle the JSON message I sent? 

There are too little info and discussion I can find related to RavenDB Queue Sink.

Thanks

**Reply by ayende** [reply]
You are getting the error from NServiceBus, not from RavenDB.
It appears that RavenDB is able to send the data successfully to the queue.
The issue is that the application on the other end (NServiceBus), doesn't like what you sent it.

**Reply by GustavXie** [reply]
No, what I am doing is I send the message to RabbitMQ and the message in the queue will be consumed by the RavenDB and make message to be a document. For this case RavenDB is a consumer

**Reply by ayende** [reply]
Given that the problem comes from NServiceBus, you'll have to check with them why that is the case.

**Reply by GustavXie** [reply]
The weird thing is it will have no issue if I send the same message twice to RabbitMQ.....  the first one may have this issue but the second one will be handled correctly without any error by NserviceBus and RavenDB can create the doc successfully based on the message from RabbitMQ

**Reply by ayende** [reply]
What error do you get? Because the error you sent did not come from RavenDB

**Reply by GustavXie** [reply]
No error I can find from RavenDB log.

**Reply by ayende** [reply]
You need to provide more information. You showed an error earlier, from NServiceBus.
Where are you getting `Could not find metadata for 'Newtonsoft.Json.Linq.JObject'` error on?

**Reply by GustavXie** [reply]
It is in the error queue with such error after I send the message to RabbitMQ

**Reply by ayende** [reply]
That message is coming from NServiceBus.

Is it possible that both NServiceBus and RavenDB compete on the same set of messages?

**Reply by GustavXie** [reply]
hmmm I have no idea, will [Queue Sink: RabbitMQ](https://ravendb.net/docs/article-page/6.0/All/server/ongoing-tasks/queue-sink/rabbit-mq-queue-sink) have such situation?

**Reply by ayende** [reply]
Let's go back to the start.
Please provide a full reproduction of this issue, so we can help.

* Setup instructions for RabbitMQ
* The actual messages you send 
* The RavenDB configuration for the sink
* Expected & actual results

**Reply by GustavXie** [reply]
Hey @ayende, I think you are right on the point: 
` Is it possible that both NServiceBus and RavenDB compete on the same set of messages?`
I think this is why it will have 50 50 chance to create doc successfully. The subsystem is monitoring the queue and it will have chance to handle the message from the queue. However I did not define the message type so it will fail anyway. 

On the other hand, I have another question about creating the Sink task. Is it possible to check whether we have created sink task in ravenDB? For now anytime I run the server, it will create a new sink task. Even I have created the same name of the sink task, it will keep creating whenever I run the server. Do you have any idea? 

Thanks

**Reply by ayende** [reply]
How are you creating the sink task right now?

**Reply by arekpalinski** [reply]
How do you create the sink task? Do you define and create it from the code or manually with the usage of RavenDB Studio?

There is `GetOngoingTaskInfoOperation()` which allows you to check the existence of a task. Unfortunately we've just found that there is a bug there for queue sinks tasks and we fixed it in:

https://github.com/ravendb/ravendb/pull/17897

**Reply by GustavXie** [reply]
I create the sink task from code. So I think I need GetOngoingTaskInfoOperation() to check the existing sink task first

**Reply by arekpalinski** [reply]
This fix will be available in RavenDB 6.0.2 which is going to be released next week. Meanwhile you can use 6.0 nightly version during the development (both the client API and the server).

**Reply by GustavXie** [reply]
Thanks a lot
